src/Accelerators/NNPA/Runtime/zDNNExtension/zDNNExtension.c

```diff
@@ -68,201 +68,378 @@ void zDNNExtensionInit() {
+            // Copy one page at a time.
+            uint64_t offsetSrc = AIU_PAGESIZE_IN_BYTES * SD3Offset;
+            uint64_t offsetDst = AIU_PAGESIZE_IN_BYTES * TD3Offset;
+            memcpy(dst + offsetDst, src + offsetSrc, AIU_PAGESIZE_IN_BYTES);
+          }
+        }
+      }
+    }
+    return;
+  }
 
-  for (uint64_t d6 = 0; d6 < D6; ++d6) {
-    for (uint64_t d5 = 0; d5 < D5; ++d5) {
-      for (uint64_t d4 = 0; d4 < D4; ++d4) {
-        uint64_t SD4Offset = d4 + D4 * (d5 + D5 * d6);
-        uint64_t TD4Offset = d4 + D4 * (d5 + D5 * d6);
-        for (uint64_t d3 = 0; d3 < D3; ++d3) {
-          uint64_t sd3 = (fromChunk ? d3 : (offset + d3));
-          uint64_t td3 = (fromChunk ? (offset + d3) : d3);
-          uint64_t SD3Offset = sd3 + SD3 * SD4Offset;
-          uint64_t TD3Offset = td3 + TD3 * TD4Offset;
-          // Copy one page at a time.
-          uint64_t offsetSrc = AIU_PAGESIZE_IN_BYTES * SD3Offset;
-          uint64_t offsetDst = AIU_PAGESIZE_IN_BYTES * TD3Offset;
-          memcpy(dst + offsetDst, src + offsetSrc, AIU_PAGESIZE_IN_BYTES);
+  // Splitting e3.
+  if (splitInfo->axis == E3) {
+    uint64_t SD4 = (fullToTile ? shapeOfFull.d4 : shapeOfTile.d4);
+    uint64_t TD4 = (fullToTile ? shapeOfTile.d4 : shapeOfFull.d4);
+    for (uint64_t d6 = 0; d6 < D6; ++d6) {
+      for (uint64_t d5 = 0; d5 < D5; ++d5) {
+        for (uint64_t d4 = 0; d4 < D4; ++d4) {
+          uint64_t sd4 = (fullToTile ? (offset + d4) : d4);
+          uint64_t td4 = (fullToTile ? d4 : (offset + d4));
+          uint64_t SD4Offset = sd4 + SD4 * (d5 + D5 * d6);
+          uint64_t TD4Offset = td4 + TD4 * (d5 + D5 * d6);
+          uint64_t copyInBytes = D3 * AIU_PAGESIZE_IN_BYTES;
+          uint64_t offsetSrc = copyInBytes * SD4Offset;
+          uint64_t offsetDst = copyInBytes * TD4Offset;
+          memcpy(dst + offsetDst, src + offsetSrc, copyInBytes);
         }
       }
     }
+    return;
+  }
+
+  // Splitting e4.
+  if (splitInfo->axis == E4) {
+    uint64_t SD6 = (fullToTile ? shapeOfFull.d6 : shapeOfTile.d6);
+    uint64_t TD6 = (fullToTile ? shapeOfTile.d6 : shapeOfFull.d6);
+    for (uint64_t d6 = 0; d6 < D6; ++d6) {
+      uint64_t sd6 = (fullToTile ? (offset + d6) : d6);
+      uint64_t td6 = (fullToTile ? d6 : (offset + d6));
+      uint64_t copyInBytes = D5 * D4 * D3 * AIU_PAGESIZE_IN_BYTES;
+      uint64_t offsetSrc = sd6 * copyInBytes;
+      uint64_t offsetDst = td6 * copyInBytes;
+      memcpy(dst + offsetDst, src + offsetSrc, copyInBytes);
+    }
+    return;
   }
-  return;
 }
 
-static void copyZTensorChunkScalar(
-    const SplitInfo *splitInfo, uint32_t chunkID, bool fromChunk) {
-  // Only support the second innermost axis in the CPU tensor at this moment.
-  // axis = 2 in the CPU tensor corresponds to dim3 in zTensor.
-  if (splitInfo->axis != 2) {
-    printf("Only support the second innermost dimension at this moment.");
+static void copyDataForTileScalar(
```

Comment:
Do you mind adding a high level comment of the difference between `CopyDataForTile` and `copyDataForTileScalar`?

Reply:
Thanks, that is super useful, I was trying to understand the why, and now it makes perfect sense.